### ISSUE TITLE
PublicClientApplication auth fails when username contains hash symbol (#)

### ISSUE THREAD

**Issue by Tim-Ford-Imprivata** [AzureAD/microsoft-authentication-library-for-java#371]
### Summary
I'm doing password authentication using the PublicClientApplication. I am finding that, if the UPN includes a hash symbol, authentication fails. 

Here is a example of pseudo code showing the issue:
```
  String username = "upn#with#hashes@example.microsoftonline.com";
  UserNamePasswordParameters parameters = 
	UserNamePasswordParameters
      .builder(SCOPE, username, password)
	  .build();
  PublicClientApplication pca = 
	PublicClientApplication.builder(clientId)
      .authority(AUTHORITY_URL)
      .build();
  pca.acquireToken(parameters);
```

Here is the resulting stack trace:
 ```
 java.lang.IllegalArgumentException: argument "content" is null
	at com.fasterxml.jackson.databind.ObjectMapper._assertNotNull(ObjectMapper.java:4737)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3435)
	at com.microsoft.aad.msal4j.JsonHelper.convertJsonToObject(JsonHelper.java:30)
	at com.microsoft.aad.msal4j.UserDiscoveryRequest.execute(UserDiscoveryRequest.java:32)
	at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.processPasswordGrant(AcquireTokenByAuthorizationGrantSupplier.java:91)
	at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.execute(AcquireTokenByAuthorizationGrantSupplier.java:43)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:59)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:17)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run$$$capture(CompletableFuture.java:1764)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1756)
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)

```
### Interpretation
It looks like the UserDiscoveryRequest.execute() gets a response with a 400 statusCode and a null body. The null body then blows up in the JsonHelper.convertJsonToObject().

Here is the code from UserDiscoveryRequest.execute():
```
  IHttpResponse response = HttpHelper.executeHttpRequest(httpRequest, requestContext, serviceBundle);

  return JsonHelper.convertJsonToObject(response.body(), UserDiscoveryResponse.class);
```

The URL used in UserDiscoveryRequest.execute() is: `https://login.microsoftonline.com/common/userrealm/upn#with#hashes@example.onmicrosoft.com?api-version=1.0`

I tried to URL encode the UPN but then the user could not be found in Azure AD.

Perhaps this is an issue with Azure AD. Still, maybe there is workaround with MSAL4J that I could use.

Or maybe the response status code could be checked before trying to parse the response. At least then the failure with the HTTP request could be reported.

Thanks for any help you can provide.

**Comment by sangonzal** [comment]
Hi @Tim-Ford-Imprivata I agree that there should be better error handling. I'll mark this issue as an enhancement and well make this change in one of the upcoming releases.

I also tried URL encoding the usernames but it doesn't seem to work. What does work is removing the hash symbol. For instance if I only have user  `test#user@test.onmicrosoft.com` in my tenant, and use `testuser@test.onmicrosoft.com` in the user discovery request, I get a valid response (even though there is no user `testuser@test.onmicrosoft.com`)

What is interesting is that I can register both `test#user@test.onmicrosoft.com` and `testuser@test.onmicrosoft.com`, and I don't know which one `https://login.microsoftonline.com/common/userrealm/testuser@sgonz.onmicrosoft.com?api-version=1.0` is referencing. 

I have reached out the the team that owns this endpoint, and will update this thread once I hear back.
